WebKit Should Log Core Animation Commits on Frame Updates

connorpobrien opened the pull request:
#### 57b86f69e2aa4425580e605dc0137bb94e511a11
<pre>
WebKit Should Log Core Animation Commits on Frame Updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=275373">https://bugs.webkit.org/show_bug.cgi?id=275373</a>
<a href="https://rdar.apple.com/129343146">rdar://129343146</a>

Reviewed by NOBODY (OOPS!).

(WIP) This commit adds CA logging for layer updates in WebKit.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeCommitTracing.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeCommitTracing.mm: Added.
(WebKit::CAFrameLogging::CAFrameLogging):
(WebKit::CAFrameLogging::~CAFrameLogging):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57b86f69e2aa4425580e605dc0137bb94e511a11

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57216 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36544 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9691 "Hash 57b86f69 for PR 29773 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60837 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7659 "Built successfully") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44168 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7849 "Hash 57b86f69 for PR 29773 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46327 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5393 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59246 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/44168 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/9691 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27187 "Found 1 new API test failure: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/44168 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/9691 "Hash 57b86f69 for PR 29773 does not build (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6664 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/44168 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/9691 "Hash 57b86f69 for PR 29773 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62517 "Built successfully") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1129 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/7849 "Hash 57b86f69 for PR 29773 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53586 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1134 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/9691 "Hash 57b86f69 for PR 29773 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53658 "2 api tests failed or timed out") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32373 "Hash 57b86f69 for PR 29773 does not build (failure)") | | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33458 "Hash 57b86f69 for PR 29773 does not build (failure)") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34543 "Hash 57b86f69 for PR 29773 does not build (failure)") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33204 "Hash 57b86f69 for PR 29773 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -41,6 +41,7 @@
 #import "WebPageProxy.h"
 #import "WebProcessProxy.h"
 #import "WindowKind.h"
+#import <QuartzCore/CATransactionPrivate.h>
```
You can't just use SPI directly like this; you need to put what you need in QuartzCoreSPI.h (follow the existing code there).

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -250,6 +251,8 @@
     LOG_WITH_STREAM(RemoteLayerTree, stream << "RemoteLayerTreeDrawingAreaProxy::commitLayerTree transaction:" << layerTreeTransaction.description());
     LOG_WITH_STREAM(RemoteLayerTree, stream << "RemoteLayerTreeDrawingAreaProxy::commitLayerTree scrolling tree:" << scrollingTreeTransaction.description());
 
+    const CFTimeInterval beginTime = CFAbsoluteTimeGetCurrent();
```
Simon suggested a C++ RAII class to handle this bookending for you, and I think he was right.

hortont424 replied:
Said class would also obviate the need for the indirection that we added in RemoteScrollingTree.

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -579,6 +579,12 @@ static bool layerEventRegionContainsPoint(CALayer *layer, CGPoint localPoint)
     });
 }
 
+void RemoteScrollingTreeMac::CAFrameLogging(CFTimeInterval beginTime, CFTimeInterval commitDeadline)
```
This method name is not great (if it's even necessary once you move this into a wrapper object).

hortont424 replied:
(if you don't factor it out, which I hope you do, I can try to help come up with a better name)

aprotyas reviewed (commented):
I think something might have gone wrong in your previous rebase, because `RemoteLayerTreeCommitTracing.h/mm` is missing from this latest commit!

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeCommitTracing.h:
```diff
@@ -0,0 +1,45 @@
+ */
+
+#pragma once
+#include <CoreFoundation/CoreFoundation.h>
+#include <QuartzCore/CALayer.h>
+#include <QuartzCore/CATransaction.h>
+
+#include <QuartzCore/CATransactionPrivate.h>
+#include <QuartzCore/QuartzCore.h>
```
You won't have to pull in these headers to declare `CAFrameLogging`

aprotyas replied:
(Also, you can try forward declaring `CAFrameToken`)

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -250,6 +252,9 @@
     LOG_WITH_STREAM(RemoteLayerTree, stream << "RemoteLayerTreeDrawingAreaProxy::commitLayerTree scrolling tree:" << scrollingTreeTransaction.description());
 
+    // Deconstructs at end of function
+    CAFrameLogging CAFrame;
```
The variable name is not ideal (style-guide wise) since we want to start with lower letters. Maybe just rename to this?

```suggestion
    CAFrameLogging frameLogging;
```

connorpobrien committed:
WebKit Should Log Core Animation Commits on Frame Updates
https://bugs.webkit.org/show_bug.cgi?id=275373
rdar://129343146

Reviewed by NOBODY (OOPS!).

(WIP) This commit adds CA logging for layer updates in WebKit.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeCommitTracing.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeCommitTracing.mm: Added.
(WebKit::CAFrameLogging::CAFrameLogging):
(WebKit::CAFrameLogging::~CAFrameLogging):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
